### PR TITLE
add Camera Auto Mount Pivoting Oblique Survey (CAMPOS)

### PR DESCRIPTION
**Pull request by igorsgcampos**
**Problem**
A solution commonly used to increase the coverage and efficiency in mapping surveys is to have multiple cameras at oblique angles to capture a wider areas per flight line, this also has potential benefits of capturing more data from facades and tall objects. However, adding cameras increase cost and complexity of the equipment, on the other hand, mapping cameras are generally capable of shooting multiple frames per second and are usually triggered about once every two seconds or less frequently in mapping missions, we could better use the hardware capabilities if we moved the camera mount between shots while decreasing the shooting distance to keep the overlap settings.

**Solution**
I decided to implement a solution that uses the already existing VEHICLE_CMD_DO_SET_CAM_TRIGG_DIST, plus some unused parameters, 4~6, to issue VEHICLE_CMD_DO_MOUNT_CONTROL between camera shots to move the camera from left to right in stages, right after it passes half the shooting distance, allowing time for mount actuation and camera exposure. To keep the frontlap the solution expects the GCS to adjust the trigger distance accordingly. The parameters usage are as follows:

- 4: number of oblique positions (this will generally be 2 or 3, but could be more, depending on desired frontlap, flying speed and hardware capabilities (how fast are the camera and mount))
- 5: the maximum angle to roll the camera left and right from the center, the solution calculates the interval angle between positions from this information and the previous parameter
- 6: the pitch angle for mounts that support pitching
https://github.com/PX4/Firmware/pull/15882

**Implementation**
I proposed the change to mavlink standard here: https://github.com/mavlink/mavlink/pull/1508
I added it to QGC here: https://github.com/mavlink/qgroundcontrol/pull/9103

**Alternatives**
Other way would be to add a new command, but since it has many aspects that are essentially the same as the CAM_TRIGG_DIST, I thought it would be better to add a few parameters to the existing command, as it would also allow it to work under Mavlink 1 with the limited number of 255 commands.

**Test data / coverage**
I tested it with the gazebo_typhoon_h480 model and checked the outcome from the video stream in QGC.

**Additional context**
Later I plan to write a paper comparing the gains in area coverage and any accuracy change from using this solution with relation to the traditional survey mode. I am feeling confident about a positive outcome.

### CODE REVIEW

File: src/drivers/camera_trigger/camera_trigger.cpp
Comment (dagar) [inline src/drivers/camera_trigger/camera_trigger.cpp:274]: I realize you're being consistent with what's already in place, but would you mind doing this as in-class member initializers? It makes it easier to ensure everything is initialized and then you don't have to fight initializing ordering.

File: src/drivers/camera_trigger/camera_trigger.cpp
Comment (dagar) [inline src/drivers/camera_trigger/camera_trigger.cpp:181]:
```suggestion
	uint32_t		_pseudo_oblique_num_poses{0};
```

File: src/drivers/camera_trigger/camera_trigger.cpp
Comment (dagar) [inline src/drivers/camera_trigger/camera_trigger.cpp:537]: You might want to take this from `vehicle_status`.

**Comment by igorsgcampos** [comment]
> > I decided to implement a solution that uses the already existing VEHICLE_CMD_DO_SET_CAM_TRIGG_DIST, plus some unused parameters, 4~6, to issue VEHICLE_CMD_DO_MOUNT_CONTROL between camera shots to move the camera from left to right in stages, right after it passes half the shooting distance, allowing time for mount actuation and camera exposure. To keep the frontlap the solution expects the GCS to adjust the trigger distance accordingly.
> 
> Technically we can do whatever we want with vehicle_command, however most of them are actually kept in sync with the corresponding mavlink command.
> 
> https://mavlink.io/en/messages/common.html#MAV_CMD_DO_MOUNT_CONTROL

Yes, I am aware of that. I was wondering if this is something we could change/add in the mavlink standard, and if so, how to do it.

File: src/drivers/camera_trigger/camera_trigger.cpp
Comment (igorsgcampos) [reply src/drivers/camera_trigger/camera_trigger.cpp:537]: I actually store and keep the target_system and target_component for the new solution when it receives a valid command.

I just added this at the test function as it was not working, these would be the standard values, and testing can happen despite receiving a valid command. Could you elaborate on how to get this information from vehicle_status?

File: src/drivers/camera_trigger/camera_trigger.cpp
Comment (igorsgcampos) [reply src/drivers/camera_trigger/camera_trigger.cpp:274]: Indeed I was following the standard in place, I wouldn't mind changing it, but I wonder what the other developers would think about having two standards in the same file.

**Comment by igorsgcampos** [comment]
> > > I decided to implement a solution that uses the already existing VEHICLE_CMD_DO_SET_CAM_TRIGG_DIST, plus some unused parameters, 4~6, to issue VEHICLE_CMD_DO_MOUNT_CONTROL between camera shots to move the camera from left to right in stages, right after it passes half the shooting distance, allowing time for mount actuation and camera exposure. To keep the frontlap the solution expects the GCS to adjust the trigger distance accordingly.
> > 
> > 
> > Technically we can do whatever we want with vehicle_command, however most of them are actually kept in sync with the corresponding mavlink command.
> > https://mavlink.io/en/messages/common.html#MAV_CMD_DO_MOUNT_CONTROL
> 
> Yes, I am aware of that. I was wondering if this is something we could change/add in the mavlink standard, and if so, how to do it.

I looked up and found the way to do it, I added a PR in mavlink standard at https://github.com/mavlink/mavlink/pull/1508, hopefully I won't have to join the dev-call, as it is 3 a.m. local time, but If necessary to further discuss it, I am willing to.

**Comment by igorsgcampos** [comment]
I would appreciate if anyone could label this as dev-call to discuss it next week.

**Comment by igorsgcampos** [comment]
I made a small video of the concept with the SITL. It is available at the following link:
https://photos.app.goo.gl/PDEye5Jx66f4f9E46

Thanks for your attention and support!

**Comment by Jaeyoung-Lim** [comment]
@igorsgcampos Have you tested/plan to test this on a real vehicle to see if it actually improves the accuracy of the map?

**Comment by hamishwillee** [comment]
@dagar @julianoes Can we get this reviewed? It is dependent on https://github.com/mavlink/mavlink/pull/1508 - but we won't merge that until this has been approved to go in (ie once we know it _definitely_ will and know when it will be merged).

**Comment by igorsgcampos** [comment]
build will break until mavlink submodule is updated with the changes from https://github.com/mavlink/mavlink/pull/1508

**Comment by igorsgcampos** [comment]
bump @dagar @julianoes

**Comment by igorsgcampos** [comment]
I plan on joining this week's dev-call to clarify any questions about this.

**Comment by Jaeyoung-Lim** [comment]
@igorsgcampos You need to rebase your PR

**Comment by Jaeyoung-Lim** [comment]
@igorsgcampos It is better to rebase, not merge: https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase

**Comment by igorsgcampos** [comment]
Damn, I just realized it was flash, not RAM that was overflowing.

**Comment by igorsgcampos** [comment]
so... apparently my latest and more aggressive attempt to reduce flash usage actually caused an increase of 8 bytes, instead of a reduction of 17 bytes...

**Comment by igorsgcampos** [comment]
I ended up concluding that this new feature is not a flash hog, the build for fmu-v2 is failing for 17-25 bytes only.even after trying to remove instructions I couldn't manage to get it to fit.

I suppose there will be flash optimization in other parts of the firmware. Eventually support for fmu-v2 will be dropped due to its limited flash.

I would appreciate if anyone could review this as it is ready, and the failed checks are not consequence of it.

**Comment by igorsgcampos** [comment]
@dagar @julianoes I rebased the PR on top of master and now it fits withing the fmuv2 flash limits, could you review this as it is ready and passing all checks now?

**Comment by igorsgcampos** [comment]
I would appreciate if anyone could add the dev-call label to this PR.

**Comment by igorsgcampos** [comment]
I added a couple flight logs from SITL as requested in the dev-call (in the first one I killed SITL before ending flight as I only wanted the images it generated to process it, the second is complete):

https://review.px4.io/plot_app?log=61c2860e-3c72-4ea3-aa29-ef046e16a9e2

https://review.px4.io/plot_app?log=eea514d0-1ec7-4f17-a471-09ea1b83e1e8

**Comment by igorsgcampos** [comment]
removed mavlink submodule commit to resolve conflict.

**Comment by igorsgcampos** [comment]
can we add this to the dev-call yet again?

**Comment by igorsgcampos** [comment]
all 94 tests were passing before I removed the mavlink submodule commit to resolve conflict, no changes made to the code, the SITL fail is not related to the PR.

**Review by LorenzMeier — APPROVED** [review]

**Pull request merged**
